### PR TITLE
feat(sc, tx): Update

### PR DESCRIPTION
**Pull request by MorganDream**
Refer to https://github.com/CityOfZion/neon-js/issues/441

### CODE REVIEW

File: packages/neon-core/src/tx/transaction/Transaction.ts [hunk 1]
```diff
@@ -289,8 +286,15 @@ export class Transaction {
   }
 
   public getScriptHashesForVerifying(): string[] {
```
Comment (snowypowers) [inline packages/neon-core/src/tx/transaction/Transaction.ts]: whats the use of this again?

File: packages/neon-core/src/sc/ScriptBuilder.ts [hunk 2]
```diff
@@ -87,7 +87,10 @@ export class ScriptBuilder extends StringStream {
     return this.emitSysCall(InteropServiceCode.SYSTEM_CONTRACT_CALL);
   }
 
-  public emitSysCall(service: InteropServiceCode) {
+  public emitSysCall(service: InteropServiceCode, ...args: any[]) {
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ScriptBuilder.ts]: tests?

File: packages/neon-core/src/tx/transaction/Transaction.ts [hunk 3]
```diff
@@ -289,8 +286,15 @@ export class Transaction {
...
-    let hashes = this.cosigners.map(cosigner => cosigner.account);
-    hashes.unshift(this.sender);
+    const hashes = this.cosigners.map(cosigner => cosigner.account);
+    if (
+      !hashes.every(hash => hashes.indexOf(hash) === hashes.lastIndexOf(hash))
+    ) {
+      throw new Error(`Duplicate Cosigner`);
```
Comment (snowypowers) [inline packages/neon-core/src/tx/transaction/Transaction.ts]: this looks like something we should be blocking on construction of transaction if this state is illegal.

File: packages/neon-core/src/tx/transaction/Transaction.ts (hunk 1 above)
Comment (MorganDream) [reply packages/neon-core/src/tx/transaction/Transaction.ts]: Get all the scripthashes that need to sign this transaction, it's used in verification. It'll be used in TransactionBuilder.

File: packages/neon-core/src/sc/ScriptBuilder.ts (hunk 2 above)
Comment (MorganDream) [reply packages/neon-core/src/sc/ScriptBuilder.ts]: Added

File: packages/neon-core/src/tx/transaction/Transaction.ts (hunk 3 above)
Comment (MorganDream) [reply packages/neon-core/src/tx/transaction/Transaction.ts]: Done

File: packages/neon-core/src/tx/transaction/Transaction.ts [hunk 4]
```diff
@@ -184,7 +183,17 @@ export class Transaction {
   }
 
   public addCosigner(...cosigners: CosignerLike[]): this {
-    this.cosigners.push(...cosigners.map(cosigner => new Cosigner(cosigner)));
+    const cosignerHashes = this.cosigners.map(cosigner => cosigner.account);
+    this.cosigners.push(
+      ...cosigners.map(cosigner => {
```
Comment (snowypowers) [inline packages/neon-core/src/tx/transaction/Transaction.ts]: `addCosigner(sameSig, sameSig)` will pass.

Comment (MorganDream) [reply packages/neon-core/src/tx/transaction/Transaction.ts]: https://github.com/CityOfZion/neon-js/blob/a30d8da56faf3c78705c5f3f84a0c6d2cead311b/packages/neon-core/__tests__/tx/transaction/Transaction.ts#L169-L183
Will not.

**Comment by hal0x2328** [comment]
200 points morgandream
100 points snowypowers for review

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/44C84C9E7582DC00411BC9D9D075F6993BA7E203FFDDE366E21DFEBD125EA580

**Pull request merged**